File: src/main/java/org/treblereel/gwt/three4g/math/Box3.java
package org.treblereel.gwt.three4g.math;

import jsinterop.annotations.JsConstructor;
import jsinterop.annotations.JsType;
import org.treblereel.gwt.three4g.core.BufferAttribute;
import org.treblereel.gwt.three4g.core.Object3D;

/**
 * Represents a box or cube in 3D space. The main purpose of this is to represent the Minimum Bounding Boxes for objects.
 *
 * @author Dmitrii Tikhomirov
 * Created by treblereel on 3/1/18.
 */
@JsType(isNative = true, namespace = "THREE")
public class Box3 implements Box {

    /**
     * Used to check whether this or derived classes are Box3s. Default is true.
     */
    public boolean isBox3;

    /**
     * Vector3 representing the lower (x, y, z) boundary of the box.
     * Default is ( + Infinity, + Infinity, + Infinity ).
     */
    public Vector3 min;

    /**
     * Vector3 representing the upper (x, y, z) boundary of the box.
     * Default is ( - Infinity, - Infinity, - Infinity ).
     */
    public Vector3 max;

    @JsConstructor
    public Box3() {

    }

    /**
     * @param min - Vector3 representing the lower (x, y, z) boundary of the box. Default is ( + Infinity, + Infinity, + Infinity ).
     * @param max - Vector3 representing the lower upper (x, y, z) boundary of the box. Default is ( - Infinity, - Infinity, - Infinity ).
     */
    public Box3(Vector3 min, Vector3 max) {

    }

    /**
     * Transforms this Box3 with the supplied matrix.
     *
     * @param matrix - The Matrix4 to apply
     * @return instance of Box3
     */
    public native Box3 applyMatrix4(Matrix4 matrix);

    /**
     * Clamps the point within the bounds of this box.
     *
     * @param point  - Vector3 to clamp.
     * @param target — the result will be copied into this Vector3.
     * @return instance of Vector3
     */
    public native Vector3 clampPoint(Vector3 point, Vector3 target);

    /**
     * @return Returns a new Vector3 with the same min and max as this one.
     */
    public native Vector3 clone();

    /**
     * @param box - Box3 to test for inclusion.
     * @return Returns true if this box includes the entirety of box. If this and box are identical,
     * * this function also returns true.
     */
    public native boolean containsBox(Box3 box);

    /**
     * @param point - Vector3 to check for inclusion.
     * @return Returns true if the specified point lies within or on the boundaries of this box.
     */
    public native boolean containsPoint(Vector3 point);

    /**
     * Copies the min and max from box to this box.
     *
     * @param box - Box3 to copy.
     * @return box instance of Box3
     */
    public native Box3 copy(Box3 box);

    /**
     * @param point - Vector3 to measure distance to.
     * @return Returns the distance from any edge of this box to the specified point. If the point lies inside of this box, the distance will be 0.
     */
    public native float distanceToPoint(Vector3 point);

    /**
     * @param box - Box to compare with this one.
     * @return Returns true if this box and box share the same lower and upper bounds.
     */
    public native boolean equals(Box3 box);

    /**
     * Expands the boundaries of this box to include object and its children, accounting for the object's, and children's, world transforms.
     *
     * @param object - Object3D to expand the box by.
     * @return box instance of Box3
     */
    public native Box3 expandByObject(Object3D object);

    /**
     * @param point - Vector3 that should be included in the box.
     * @return Expands the boundaries of this box to include point.
     */
    public native Box3 expandByPoint(Vector3 point);

    /**
     * Expands each dimension of the box by scalar. If negative, the dimensions of the box will be contracted.
     *
     * @param scalar - Distance to expand the box by.
     * @return box instance of Box3
     */
    public native Box3 expandByScalar(float scalar);

    /**
     * Expands this box equilaterally by vector. The width of this box will be expanded by the x component of vector in both directions. The height of this box will be expanded by the y component of vector in both directions.
     *
     * @param vector - Vector3 to expand the box by.
     * @return box instance of Box3
     */
    public native Box3 expandByVector(Vector3 vector);

    /**
     * Gets a Sphere that bounds the box.
     *
     * @param target — the result will be copied into this Sphere.
     * @return Gets a Sphere that bounds the box.
     */
    public native Sphere getBoundingSphere(Sphere target);

    /**
     * @param target — the result will be copied into this Vector3.
     * @return Returns the center point of the box as a Vector3.
     */
    public native Vector3 getCenter(Vector3 target);

    /**
     * @param point  - Vector3.
     * @param target — the result will be copied into this Vector3.
     * @return Returns a point as a proportion of this box's width and height.
     */
    public native Vector3 getParameter(Vector3 point, Vector3 target);

    /**
     * @param target — the result will be copied into this Vector3.
     * @return Returns the width and height of this box.
     */
    public native Vector3 getSize(Vector3 target);

    /**
     * @param box - Box to intersect with.
     * @return Returns the intersection of this and box, setting the upper bound of this box to the lesser of the two boxes' upper bounds and the lower bound of this box to the greater of the two boxes' lower bounds.
     */
    public native Box3 intersect(Box3 box);

    /**
     * Determines whether or not this box intersects box.
     *
     * @param box - Box to check for intersection against.
     * @return true, if this box intersects box
     */
    public native boolean intersectsBox(Box3 box);

    /**
     * Determines whether or not this box intersects plane.
     *
     * @param plane - Plane to check for intersection against.
     * @return true, if this plane intersects box
     */
    public native boolean intersectsPlane(Plane plane);

    /**
     * Determines whether or not this box intersects sphere.
     *
     * @param sphere - Sphere to check for intersection against.
     * @return true, if this box intersects sphere
     */
    public native boolean intersectsSphere(Sphere sphere);

    /**
     * Determines whether or not this box intersects triangle.
     *
     * @param triangle - Triangle to check for intersection against.
     * @return true, if this box intersects box
     */
    public native boolean intersectsTriangle(Triangle triangle);

    /**
     * Note that a box with equal lower and upper bounds still includes one point, the one both bounds share.
     *
     * @return Returns true if this box includes zero points within its bounds.
     */
    public native boolean isEmpty();

    /**
     * Makes this box empty.
     *
     * @return empty Box3
     */
    public native Box3 makeEmpty();

    /**
     * Sets the lower and upper (x, y, z) boundaries of this box.
     *
     * @param min - Vector3 representing the lower (x, y, z) boundary of the box.
     * @param max - Vector3 representing the lower upper (x, y, z) boundary of the box.
     * @return instance of Box3
     */
    public native Box3 set(Vector3 min, Vector3 max);

    /**
     * Sets the upper and lower bounds of this box to include all of the data in array.
     *
     * @param array -- An array of position data that the resulting box will envelop.
     * @return instance of Box3
     */
    public native Box3 setFromArray(int[] array);

    /**
     * Sets the upper and lower bounds of this box to include all of the data in attribute.
     *
     * @param attribute - A buffer attribute of position data that the resulting box will envelop.
     * @return instance of Box3
     */
    public native Box3 setFromBufferAttribute(BufferAttribute attribute);

    /**
     * Centers this box on center and sets this box's width and height to the values specified in size.
     *
     * @param center - Desired center position of the box (Vector3).
     * @param size   - Desired x, y and z dimensions of the box (Vector3).
     * @return instance of Box3
     */
    public native Box3 setFromCenterAndSize(Vector3 center, Vector3 size);


    /**
     * Computes the world-axis-aligned bounding box of an Object3D (including its children), accounting for the object's, and children's, world transforms.
     *
     * @param object - Object3D to compute the bounding box of.
     * @return instance of Box3
     */
    public native Box3 setFromObject(Object3D object);


    /**
     * Sets the upper and lower bounds of this box to include all of the points in points.
     *
     * @param points - Array of Vector3s that the resulting box will contain.
     * @return instance of Box3
     */
    public native Box3 setFromPoints(Vector3... points);

    /**
     * Adds offset to both the upper and lower bounds of this box, effectively moving this box offset units in 3D space.     *
     *
     * @param offset - Direction and distance of offset.
     * @return instance of Box3
     */
    public native Box3 translate(Vector3 offset);

    /**
     * Unions this box with box, setting the upper bound of this box to the greater of the two boxes' upper bounds and the lower bound of this box to the lesser of the two boxes' lower bounds.
     *
     * @param box - Box that will be unioned with this box.
     * @return instance of Box3
     */
    public native Box3 union(Box3 box);
}

File: core/src/main/resources/org/treblereel/gwt/three4g/math/Box3.js
import { Vector3 } from './Vector3.js';

import { Matrix4 } from './Matrix4.js';

//closure shim
import { BufferAttributeInterface } from '../closure/core/BufferAttributeInterface.js';
import { TriangleInterface } from '../closure/math/TriangleInterface.js';
import { Box3Interface } from '../closure/math/Box3Interface.js';
import { Object3DInterface } from '../closure/core/Object3DInterface.js';

/**
 * @implements {Box3Interface}
 */
class Box3 {

	/**
	 * 
	 * @param {Vector3} min 
	 * @param {Vector3} max 
	 */
	constructor( min = new Vector3( + Infinity, + Infinity, + Infinity ), max = new Vector3( - Infinity, - Infinity, - Infinity ) ) {

		/** @type {Vector3} */
		this.min = min;
		/** @type {Vector3} */
		this.max = max;

	}

	/**
	 * 
	 * @param {Vector3} min 
	 * @param {Vector3} max 
	 * @return {Box3}
	 */
	set( min, max ) {

		this.min.copy( min );
		this.max.copy( max );

		return this;

	}

	/**
	 * 
	 * @param {Array<number>} array 
	 * @return {Box3}
	 */
	setFromArray( array ) {

		let minX = + Infinity;
		let minY = + Infinity;
		let minZ = + Infinity;

		let maxX = - Infinity;
		let maxY = - Infinity;
		let maxZ = - Infinity;

		for ( let i = 0, l = array.length; i < l; i += 3 ) {

			const x = array[ i ];
			const y = array[ i + 1 ];
			const z = array[ i + 2 ];

			if ( x < minX ) minX = x;
			if ( y < minY ) minY = y;
			if ( z < minZ ) minZ = z;

			if ( x > maxX ) maxX = x;
			if ( y > maxY ) maxY = y;
			if ( z > maxZ ) maxZ = z;

		}

		this.min.set( minX, minY, minZ );
		this.max.set( maxX, maxY, maxZ );

		return this;

	}

	/**
	 * @param {BufferAttributeInterface} attribute 
	 */
	setFromBufferAttribute( attribute ) {

		let minX = + Infinity;
		let minY = + Infinity;
		let minZ = + Infinity;

		let maxX = - Infinity;
		let maxY = - Infinity;
		let maxZ = - Infinity;

		for ( let i = 0, l = attribute.count; i < l; i ++ ) {

			const x = attribute.getX( i );
			const y = attribute.getY( i );
			const z = attribute.getZ( i );

			if ( x < minX ) minX = x;
			if ( y < minY ) minY = y;
			if ( z < minZ ) minZ = z;

			if ( x > maxX ) maxX = x;
			if ( y > maxY ) maxY = y;
			if ( z > maxZ ) maxZ = z;

		}

		this.min.set( minX, minY, minZ );
		this.max.set( maxX, maxY, maxZ );

		return this;

	}

	/**
	 * @param {Array<Vector3>} points 
	 * @return {Box3}
	 */
	setFromPoints( points ) {

		this.makeEmpty();

		for ( let i = 0, il = points.length; i < il; i ++ ) {

			this.expandByPoint( points[ i ] );

		}

		return this;

	}

	/**
	 * @param {Vector3} center 
	 * @param {Vector3} size 
	 * @return {Box3}
	 */
	setFromCenterAndSize( center, size ) {

		const halfSize = _vector.copy( size ).multiplyScalar( 0.5 );

		this.min.copy( center ).sub( halfSize );
		this.max.copy( center ).add( halfSize );

		return this;

	}

	/**
	 * @param {Object3DInterface} object 
	 * @return {Box3}
	 */
	setFromObject( object ) {

		this.makeEmpty();

		return this.expandByObject( object );

	}

	/**
	 * @return {Box3}
	 */
	clone() {

		return new this.constructor().copy( this );

	}

	/**
	 * @param {Box3} box 
	 * @return {Box3}
	 */
	copy( box ) {

		this.min.copy( box.min );
		this.max.copy( box.max );

		return this;

	}

	/**
	 * 
	 * @return {Box3}
	 */
	makeEmpty() {

		this.min.x = this.min.y = this.min.z = + Infinity;
		this.max.x = this.max.y = this.max.z = - Infinity;

		return this;

	}

	/**
	 * 
	 * @return {boolean}
	 */
	isEmpty() {

		// this is a more robust check for empty than ( volume <= 0 ) because volume can get positive with two negative axes

		return ( this.max.x < this.min.x ) || ( this.max.y < this.min.y ) || ( this.max.z < this.min.z );

	}

	/**
	 * 
	 * @param {Vector3} target 
	 * @return {Vector3}
	 */
	getCenter( target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Box3: .getCenter() target is now required' );
			target = new Vector3();

		}

		return this.isEmpty() ? target.set( 0, 0, 0 ) : target.addVectors( this.min, this.max ).multiplyScalar( 0.5 );

	}

	/**
	 * 
	 * @param {Vector3} target 
	 * @return {Vector3}
	 */
	getSize( target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Box3: .getSize() target is now required' );
			target = new Vector3();

		}

		return this.isEmpty() ? target.set( 0, 0, 0 ) : target.subVectors( this.max, this.min );

	}

	/**
	 * 
	 * @param {Vector3} point 
	 * @return {Box3}
	 */
	expandByPoint( point ) {

		this.min.min( point );
		this.max.max( point );

		return this;

	}

	/**
	 * 
	 * @param {Vector3} vector 
	 * @return {Box3}
	 */
	expandByVector( vector ) {

		this.min.sub( vector );
		this.max.add( vector );

		return this;

	}

	/**
	 * 
	 * @param {number} scalar 
	 * @return {Box3}
	 */
	expandByScalar( scalar ) {

		this.min.addScalar( - scalar );
		this.max.addScalar( scalar );

		return this;

	}

	/**
	 * @param {Object3DInterface} object
 	 * @return {Box3}
	 */
	expandByObject( object ) {

		// Computes the world-axis-aligned bounding box of an object (including its children),
		// accounting for both the object's, and children's, world transforms

		object.updateWorldMatrix( false, false );

		const geometry = object.geometry;

		if ( geometry !== undefined ) {

			if ( geometry.boundingBox === null ) {

				geometry.computeBoundingBox();

			}

			_box.copy( geometry.boundingBox );
			_box.applyMatrix4( object.matrixWorld );

			this.union( _box );

		}

		const children = object.children;

		for ( let i = 0, l = children.length; i < l; i ++ ) {

			this.expandByObject( children[ i ] );

		}

		return this;

	}

	/**
	 * 
	 * @param {Vector3} point 
	 * @return {boolean}
	 */
	containsPoint( point ) {

		return point.x < this.min.x || point.x > this.max.x ||
			point.y < this.min.y || point.y > this.max.y ||
			point.z < this.min.z || point.z > this.max.z ? false : true;

	}

	/**
	 * 
	 * @param {Box3} box 
	 * @return {boolean}
	 */
	containsBox( box ) {

		return this.min.x <= box.min.x && box.max.x <= this.max.x &&
			this.min.y <= box.min.y && box.max.y <= this.max.y &&
			this.min.z <= box.min.z && box.max.z <= this.max.z;

	}

	/**
	 * 
	 * @param {Vector3} point 
	 * @param {Vector3} target 
	 * @return {Vector3}
	 */
	getParameter( point, target ) {

		// This can potentially have a divide by zero if the box
		// has a size dimension of 0.

		if ( target === undefined ) {

			console.warn( 'THREE.Box3: .getParameter() target is now required' );
			target = new Vector3();

		}

		return target.set(
			( point.x - this.min.x ) / ( this.max.x - this.min.x ),
			( point.y - this.min.y ) / ( this.max.y - this.min.y ),
			( point.z - this.min.z ) / ( this.max.z - this.min.z )
		);

	}

	/**
	 * 
	 * @param {Box3} box 
	 * @return {boolean}
	 */
	intersectsBox( box ) {

		// using 6 splitting planes to rule out intersections.
		return box.max.x < this.min.x || box.min.x > this.max.x ||
			box.max.y < this.min.y || box.min.y > this.max.y ||
			box.max.z < this.min.z || box.min.z > this.max.z ? false : true;

	}

	/**
	 * @param {{ center: Vector3, radius : number}} sphere 
	 * @return {boolean} 
	 */
	intersectsSphere( sphere ) {

		// Find the point on the AABB closest to the sphere center.
		this.clampPoint( sphere.center, _vector );

		// If that point is inside the sphere, the AABB and sphere intersect.
		return _vector.distanceToSquared( sphere.center ) <= ( sphere.radius * sphere.radius );

	}

	/**
	 * 
	 * @param { {normal : Vector3, constant: number}} plane 
	 * @return {boolean}
	 */
	intersectsPlane( plane ) {

		// We compute the minimum and maximum dot product values. If those values
		// are on the same side (back or front) of the plane, then there is no intersection.

		let min, max;

		if ( plane.normal.x > 0 ) {

			min = plane.normal.x * this.min.x;
			max = plane.normal.x * this.max.x;

		} else {

			min = plane.normal.x * this.max.x;
			max = plane.normal.x * this.min.x;

		}

		if ( plane.normal.y > 0 ) {

			min += plane.normal.y * this.min.y;
			max += plane.normal.y * this.max.y;

		} else {

			min += plane.normal.y * this.max.y;
			max += plane.normal.y * this.min.y;

		}

		if ( plane.normal.z > 0 ) {

			min += plane.normal.z * this.min.z;
			max += plane.normal.z * this.max.z;

		} else {

			min += plane.normal.z * this.max.z;
			max += plane.normal.z * this.min.z;

		}

		return ( min <= - plane.constant && max >= - plane.constant );

	}

	/**
	 * @param {TriangleInterface} triangle 
	 * @return {boolean}
	 */
	intersectsTriangle( triangle ) {

		if ( this.isEmpty() ) {

			return false;

		}

		// compute box center and extents
		this.getCenter( _center );
		_extents.subVectors( this.max, _center );

		// translate triangle to aabb origin
		_v0.subVectors( triangle.a, _center );
		_v1.subVectors( triangle.b, _center );
		_v2.subVectors( triangle.c, _center );

		// compute edge vectors for triangle
		_f0.subVectors( _v1, _v0 );
		_f1.subVectors( _v2, _v1 );
		_f2.subVectors( _v0, _v2 );

		// test against axes that are given by cross product combinations of the edges of the triangle and the edges of the aabb
		// make an axis testing of each of the 3 sides of the aabb against each of the 3 sides of the triangle = 9 axis of separation
		// axis_ij = u_i x f_j (u0, u1, u2 = face normals of aabb = x,y,z axes vectors since aabb is axis aligned)
		let axes = [
			0, - _f0.z, _f0.y, 0, - _f1.z, _f1.y, 0, - _f2.z, _f2.y,
			_f0.z, 0, - _f0.x, _f1.z, 0, - _f1.x, _f2.z, 0, - _f2.x,
			- _f0.y, _f0.x, 0, - _f1.y, _f1.x, 0, - _f2.y, _f2.x, 0
		];
		if ( ! satForAxes( axes, _v0, _v1, _v2, _extents ) ) {

			return false;

		}

		// test 3 face normals from the aabb
		axes = [ 1, 0, 0, 0, 1, 0, 0, 0, 1 ];
		if ( ! satForAxes( axes, _v0, _v1, _v2, _extents ) ) {

			return false;

		}

		// finally testing the face normal of the triangle
		// use already existing triangle edge vectors here
		_triangleNormal.crossVectors( _f0, _f1 );
		axes = [ _triangleNormal.x, _triangleNormal.y, _triangleNormal.z ];

		return satForAxes( axes, _v0, _v1, _v2, _extents );

	}

	/**
	 * @param {Vector3} point 
	 * @param {Vector3=} target
	 * @return {Vector3} 
	 */
	clampPoint( point, target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Box3: .clampPoint() target is now required' );
			target = new Vector3();

		}

		return target.copy( point ).clamp( this.min, this.max );

	}

	/**
	 * @param {Vector3} point 
	 * @return {number} 
	 */
	distanceToPoint( point ) {

		const clampedPoint = _vector.copy( point ).clamp( this.min, this.max );

		return clampedPoint.sub( point ).length();

	}

	/**
	 * @param {{ center: Vector3, radius : number}=} target 
	 */
	getBoundingSphere( target ) {

		if ( target === undefined ) {

			console.error( 'THREE.Box3: .getBoundingSphere() target is now required' );
			//target = new Sphere(); // removed to avoid cyclic dependency

		}

		this.getCenter( target.center );

		target.radius = this.getSize( _vector ).length() * 0.5;

		return target;

	}

	/**
	 * 
	 * @param {Box3} box
	 * @return {Box3} 
	 */
	intersect( box ) {

		this.min.max( box.min );
		this.max.min( box.max );

		// ensure that if there is no overlap, the result is fully empty, not slightly empty with non-inf/+inf values that will cause subsequence intersects to erroneously return valid values.
		if ( this.isEmpty() ) this.makeEmpty();

		return this;

	}

	/**
	 * 
	 * @param {Box3} box
	 * @return {Box3} 
	 */
	union( box ) {

		this.min.min( box.min );
		this.max.max( box.max );

		return this;

	}

	/**
	 * @suppress {checkTypes} 
	 * @param {Matrix4} matrix 
	 * @return {Box3}
	 */
	applyMatrix4( matrix ) {

		// transform of empty box is an empty box.
		if ( this.isEmpty() ) return this;

		// NOTE: I am using a binary pattern to specify all 2^3 combinations below
		_points[ 0 ].set( this.min.x, this.min.y, this.min.z ).applyMatrix4( matrix ); // 000
		_points[ 1 ].set( this.min.x, this.min.y, this.max.z ).applyMatrix4( matrix ); // 001
		_points[ 2 ].set( this.min.x, this.max.y, this.min.z ).applyMatrix4( matrix ); // 010
		_points[ 3 ].set( this.min.x, this.max.y, this.max.z ).applyMatrix4( matrix ); // 011
		_points[ 4 ].set( this.max.x, this.min.y, this.min.z ).applyMatrix4( matrix ); // 100
		_points[ 5 ].set( this.max.x, this.min.y, this.max.z ).applyMatrix4( matrix ); // 101
		_points[ 6 ].set( this.max.x, this.max.y, this.min.z ).applyMatrix4( matrix ); // 110
		_points[ 7 ].set( this.max.x, this.max.y, this.max.z ).applyMatrix4( matrix ); // 111

		this.setFromPoints( _points );

		return this;

	}

	/**
	 * 
	 * @param {Vector3} offset 
	 * @return {Box3}
	 */
	translate( offset ) {

		this.min.add( offset );
		this.max.add( offset );

		return this;

	}

	/**
	 * @param {Box3} box
	 * @return {boolean} 
	 */
	equals( box ) {

		return box.min.equals( this.min ) && box.max.equals( this.max );

	}

}

Box3.prototype.isBox3 = true;

/** @type {Array<Vector3>} */
const _points = [
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3(),
	/*@__PURE__*/ new Vector3()
];

const _vector = /*@__PURE__*/ new Vector3();

const _box = /*@__PURE__*/ new Box3();

// triangle centered vertices

const _v0 = /*@__PURE__*/ new Vector3();
const _v1 = /*@__PURE__*/ new Vector3();
const _v2 = /*@__PURE__*/ new Vector3();

// triangle edge vectors

const _f0 = /*@__PURE__*/ new Vector3();
const _f1 = /*@__PURE__*/ new Vector3();
const _f2 = /*@__PURE__*/ new Vector3();

const _center = /*@__PURE__*/ new Vector3();
const _extents = /*@__PURE__*/ new Vector3();
const _triangleNormal = /*@__PURE__*/ new Vector3();
const _testAxis = /*@__PURE__*/ new Vector3();

/**
 * @param {Array<number>} axes 
 * @param {Vector3} v0 
 * @param {Vector3} v1 
 * @param {Vector3} v2 
 * @param {Vector3} extents 
 * @return {boolean}
 */
function satForAxes( axes, v0, v1, v2, extents ) {

	for ( let i = 0, j = axes.length - 3; i <= j; i += 3 ) {

		_testAxis.fromArray( axes, i );
		// project the aabb onto the seperating axis
		const r = extents.x * Math.abs( _testAxis.x ) + extents.y * Math.abs( _testAxis.y ) + extents.z * Math.abs( _testAxis.z );
		// project all 3 vertices of the triangle onto the seperating axis
		const p0 = v0.dot( _testAxis );
		const p1 = v1.dot( _testAxis );
		const p2 = v2.dot( _testAxis );
		// actual test, basically see if either of the most extreme of the triangle points intersects r
		if ( Math.max( - Math.max( p0, p1, p2 ), Math.min( p0, p1, p2 ) ) > r ) {

			// points of the projected triangle are outside the projected half-length of the aabb
			// the axis is seperating and we can exit
			return false;

		}

	}

	return true;

}

export { Box3 };
